docs/design_proposals/artifact-dependencies.md

```diff
@@ -0,0 +1,356 @@
+
+>*Note: Omitted some details for brevity, assuming reader's familiarity with the skaffold [codebase](https://github.com/GoogleContainerTools/skaffold).*
+
+## Config schema
+
+We introduce an `ArtifactDependency` slice within `Artifact` in `config.go`
+```go
+type ArtifactDependency struct {
+	ImageName string `yaml:"image" yamltags:"required"`
+	Alias string `yaml:"alias,omitempty"`
+}
+```
+This allows us to define a build stanza like below where image `leeroy-app` requires image `simple-go-app`:
+```yaml
+build:
+ artifacts:
+   - image: simple-go-app
+   - image: leeroy-app
+     requires:
+       - image: simple-go-app
+         alias: BASE
+```
+Alias is a token that will be replaced with the image reference in the builder definition files. If no value is provided for `alias` then it defaults to the value of `image`.
+
+## Config validation
+
+We add two new validations to the [validation](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/schema/validation/validation.go#L37) package after the introduction of artifact dependencies:
+- Cyclic references among artifacts.
+  - We cannot have image `A` depend on image `B` depend on image `C` depend on image `A`.
+  - We run a simple depth first search cycle detection algorithm treating our `Artifact` slice like a directed graph- image `A` depending on image `B` implies a directed edge from `A` to `B`.
+- Unique artifact aliases.
+  - We ensure that within *each* artifact dependency slice the aliases are unique. 
+
+## Referencing dependencies
+
+### Docker builder
+
+The `docker` builder will use the `alias` of an `ArtifactDependency` as a build argument key.
+ 
+ ```yaml
+ build:
+  artifacts:
+    - image: simple-go-app
+    - image: leeroy-app
+      requires:
+        - image: simple-go-app
+          alias: BASE
+ ```
+
+Here `BASE=gcr.io/X/simple-go-app:<tag>@sha:<sha>` is passed as a `--build-arg` (or a buildKit parameter) when building `leeroy-app`.
+
+### Custom builder
+
+The `custom` builder will be supplied each `ArtifactDependency`'s `alias` and image reference as environment variables keyed on `alias`. So they can be easily referenced in user-defined build definitions.
+
+### Buildpacks builder
+
+Buildpacks supports overriding the run-image and the builder-image in its current schema. We extend this to allow specifying `ArtifactDependency` aliases as the value for the `runImage` and `builder` fields.
+
+```yaml
+build:
+  artifacts:
+  - image: skaffold-buildpacks
+    buildpacks:
+      builder: "{{ .BUILDER_IMAGE }}"
+      runImage: "{{ .RUN_IMAGE }}"
+    requires:
+      - image: builder-image
+        alias: BUILDER_IMAGE
+      - image: run-image
+        alias: RUN_IMAGE
+```
+
+If there are any additional images in the `required` section it only enforces that they get built prior to the current image. However, the buildpacks builder cannot really reference them in any other way.
```

Comment:
I've been thinking of our chat yesterday and what [@tejal29 suggested above](https://github.com/GoogleContainerTools/skaffold/pull/4794/files#r491234437), and I now think that Buildpacks shouldn't have these additional `requires` block, since the `builder` and `runImage` are sufficient and encode implicit dependencies.  We're repeating ourselves in the example above, and it doesn't seem to buy us anything.

Let's see if we need build ordering (implicit or explicit) and encode it then.

Reply:
It'd be useful to think about this design document with the context that my primary objective is to implement `modules` piece by piece. 

If we are saying that only the docker or custom builder can defined required modules (which could be bundles of multiple builds) then it makes sense to only implement `required` sections for those builders only. Otherwise it'll just complicate the implementation down the line; this feature by itself might only live for a couple of weeks. 

The `required` clause will migrate to the `module` struct soon enough which will be one level above `build`. As such it doesn't seem worthwhile to push it even lower into individual builder structs only to have to refactor again.

With `modules` and `hooks` skaffold is slowly transforming into a general orchestration framework. I think it's a worthwhile tradeoff to have some redundancies in schema to get conformity across whatever builders and deployers and hooks any random module might define.

Reply:
TLDR; for this buildpacks example what would the schema look like if `builder-image` was in a skaffold module in another repo?

Reply:
Good point on the modules.  I'm hoping we don't use Go templates for this; I'd rewrite your example as:
```
build:
  artifacts:
  - image: builder-image
  - image: skaffold-buildpacks
    buildpacks:
      builder: builder-image
    requires:
      - image: builder-image
        module: 
           path: ../builder/ 
```